nessie/run.go

```diff
@@ -0,0 +1,124 @@
+package nessie
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/thanhpk/randstr"
+	"github.com/treeverse/lakefs/api"
+	"github.com/treeverse/lakefs/api/gen/client"
+	"github.com/treeverse/lakefs/api/gen/client/setup"
+	"github.com/treeverse/lakefs/api/gen/models"
+	"github.com/treeverse/lakefs/logging"
+)
+
+// Config contains configuration needed for running the tests
+type Config struct {
+	// BaseURL is the base address of the lakeFS endpoint
+	BaseURL string
+
+	// BucketPath is the full path to the s3 bucket
+	BucketPath string
+}
+
+// Run runs system tests and reports on failures
+func Run(ctx context.Context, config Config) error {
+	// initialize the env/repo
+	logger := logging.FromContext(ctx)
+	bucketPath := fmt.Sprintf("%s/%s", config.BucketPath, time.Now().UTC().Format("2006/01/02/150405"))
+	logger.WithField("bucketPath", bucketPath).Info("Starting nessie run")
+
+	cl := client.NewHTTPClientWithConfig(strfmt.Default, &client.TransportConfig{
+		Host:     config.BaseURL,
+		BasePath: client.DefaultBasePath,
+		Schemes:  []string{"http"},
+	})
+
+	expBackoff := backoff.NewExponentialBackOff()
+	if err := backoff.Retry(func() error {
+		_, err := cl.HealthCheck.HealthCheck(nil)
+		if err != nil {
+			return err
+		}
+		return nil
+	}, expBackoff); err != nil {
+		return fmt.Errorf("api endpoint isn't ready: %w", err)
+	}
+
+	adminUserName := "nessie"
+	res, err := cl.Setup.SetupLakeFS(&setup.SetupLakeFSParams{
+		User: &models.Setup{
+			DisplayName: &adminUserName,
+		},
+		Context: ctx,
+	})
+	if err != nil {
+		return fmt.Errorf("lakefs setup request failed: %w", err)
+	}
+
+	logger.WithField("accessKeyID", res.Payload.AccessKeyID).
+		WithField("accessSecretKey", res.Payload.AccessSecretKey).
+		Info("Cluster setup successfully")
+
+	c, err := api.NewClient(fmt.Sprintf("http://%s/%s", config.BaseURL, client.DefaultBasePath), res.Payload.AccessKeyID, res.Payload.AccessSecretKey)
+	if err != nil {
+		return fmt.Errorf("failed to setup client: %w", err)
+	}
+
+	if err := c.CreateRepository(ctx, &models.RepositoryCreation{
+		DefaultBranch:    masterBranch,
+		ID:               swag.String(repo),
+		StorageNamespace: &bucketPath,
+	}); err != nil {
+		return fmt.Errorf("failed to create repo: %w", err)
+	}
+	logger.WithField("repo", repo).Info("Created first repository")
+
+	// TODO(itai): run the tests in parallel
+	if err := singleCommit(ctx, c); err != nil {
+		return err
+	}
+
+	logger.Info("All tests finished successfully")
+	return nil
```

Comment:
less of an issue in CI, but locally, if this was a Go test case, you'd get the benefit of [t.Cleanup](https://golang.org/pkg/testing/#T.Cleanup)